controllers/csi.go

```diff
@@ -0,0 +1,689 @@
+		{Name: "plugin-dir", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: pluginPath, Type: &hostPathDirectoryorCreate}}},
+		{Name: "csi-plugins-dir", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: fmt.Sprintf("%s/plugins/", defaultKubeletDirPath), Type: &hostPathDirectory}}},
+		{Name: "registration-dir", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: fmt.Sprintf("%s/plugins_registry/", defaultKubeletDirPath), Type: &hostPathDirectory}}},
+		{Name: "pods-mount-dir", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: fmt.Sprintf("%s/pods", defaultKubeletDirPath), Type: &hostPathDirectory}}},
+		{Name: "oidc-token", VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				Sources: []corev1.VolumeProjection{
+					{
+						ServiceAccountToken: &corev1.ServiceAccountTokenProjection{
+							Path:              "oidc-token",
+							ExpirationSeconds: pointer.Int64Ptr(3600),
+							Audience:          "ceph-csi-kms",
+						},
+					},
+				},
+			},
+		}},
+		// No support for custom ceph.conf yet
+	}
+
+	labels := map[string]string{
+		"app": name,
+	}
+	s := sideCarContainer{
+		namespace:      namespace,
+		clusterVersion: clusterVersion,
+	}
+	rbdPluginContainer := s.getCephCsiContainer("rbd", driverName, false)
+	// set security context for cephfs plugin which is only required when
+	// running as daemonset
+	rbdPluginContainer.SecurityContext = &corev1.SecurityContext{
+		Privileged: pointer.BoolPtr(true),
+		Capabilities: &corev1.Capabilities{
+			Add: []corev1.Capability{"SYS_ADMIN"},
+		},
+		AllowPrivilegeEscalation: pointer.BoolPtr(true),
+	}
+	// get all containers that are part of csi controller deployment
+	containers := []corev1.Container{
+		*s.getDriverRegistrarContainer(fmt.Sprintf("%s/csi.sock", pluginPath)),
+		*rbdPluginContainer,
+		// not includuing below sidecar yet as they are not default
+		// liveness probe
+	}
+
+	pluginDaemonSet := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: namespace,
+					Labels:    labels,
+				},
+				Spec: corev1.PodSpec{
+					HostNetwork:        true, // Need to run with host networking for map/mount operation
+					HostPID:            true,
+					DNSPolicy:          corev1.DNSClusterFirstWithHostNet,
+					Containers:         containers,
+					ServiceAccountName: "csi-rbd-plugin-sa",
+					PriorityClassName:  "system-node-critical",
+					Volumes:            volumes,
+				},
+			},
+		},
+	}
+
+	return pluginDaemonSet
+
+}
+
+type sideCarContainer struct {
+	namespace      string
+	clusterVersion string
+}
```

Comment:
...why exactly is this being done? Given that it only has two fields, and basically nothing is done with them other than getting their values, all the `get...()`  functions could just be top-level functions with the two string values as parameters. If there was a more compelling reason to have a struct, like more than just returning fixed resource definitions, I'd consider it more appropriate. The functions themselves are fine.

Reply:
the namespace is used in sidecar functions to set the container argument `fmt.Sprintf("--leader-election-namespace=%s", s.namespace)`. I kept it as a struct as we might add more parameters and avoid adding more input to the function signature.

Reply:
Good enough for now!